src/main/scala/beam/agentsim/infrastructure/ChargingNetworkManager.scala

```diff
@@ -0,0 +1,477 @@
+      scheduler ! ScheduleTrigger(ChargingTimeOutTrigger(endTime, vehicle.id, vehicleManager), self)
+  }
+
+  /**
+    * Disconnect the vehicle
+    * @param tick current time
+    * @param chargingVehicle charging vehicle information
+    * @param currentSenderMaybe the current actor who sent the message
+    */
+  private def handleEndCharging(
+    tick: Int,
+    chargingVehicle: ChargingVehicle,
+    currentSenderMaybe: Option[ActorRef] = None
+  ): Unit = {
+    val ChargingVehicle(vehicle, vehicleManager, _, station, _, _) = chargingVehicle
+    val chargingNetwork = chargingNetworkMap(vehicleManager)
+    if (chargingNetwork.disconnectVehicle(vehicle)) {
+      handleRefueling(chargingVehicle)
+      handleEndChargingHelper(tick, chargingVehicle)
+      vehicle.disconnectFromChargingPoint()
+      vehicle.stall match {
+        case Some(stall) =>
+          parkingManager ! ReleaseParkingStall(stall.parkingZoneId, stall.tazId)
+          vehicle.unsetParkingStall()
+        case None =>
+          log.error(s"Vehicle $vehicle has no stall while ending charging event")
+      }
+      currentSenderMaybe.foreach(_ ! EndingRefuelSession(tick, vehicle.id))
+    } else {
+      log.debug(s"Vehicle $vehicle is already disconnected at $tick")
+    }
+    chargingNetwork.processWaitingLine(station).foreach(handleStartCharging(tick, _))
+  }
+
+  /**
+    * Refuel the vehicle using last charging session and collect the corresponding load
+    * @param chargingVehicle vehicle charging information
+    */
+  private def handleRefueling(chargingVehicle: ChargingVehicle): Unit = {
+    chargingVehicle.latestChargingCycle.foreach { cycle =>
+      val vehicle = chargingVehicle.vehicle
+      vehicle.addFuel(cycle.energy)
+      collectObservedLoadInKW(chargingVehicle, cycle)
+      log.debug(s"Charging vehicle $vehicle. Stall ${vehicle.stall}. Provided energy of = ${cycle.energy} J")
+    }
+  }
+
+  /**
+    * process the event ChargingPlugInEvent
+    * @param currentTick current time
+    * @param chargingVehicle vehicle charging information
+    */
+  private def handleStartChargingHelper(currentTick: Int, chargingVehicle: ChargingVehicle): Unit = {
+    import chargingVehicle._
+    val chargingPlugInEvent = new ChargingPlugInEvent(
+      tick = currentTick,
+      stall = stall,
+      locationWGS = geo.utm2Wgs(stall.locationUTM),
+      vehId = vehicle.id,
+      primaryFuelLevel = vehicle.primaryFuelLevelInJoules,
+      secondaryFuelLevel = Some(vehicle.secondaryFuelLevelInJoules)
+    )
+    log.debug(s"ChargingPlugInEvent: $chargingPlugInEvent")
+    beamServices.matsimServices.getEvents.processEvent(chargingPlugInEvent)
+  }
+
+  /**
+    * Calculates the duration of the refuel session, the provided energy and throws corresponding events
+    * @param currentTick current time
+    * @param chargingVehicle vehicle charging information
+    */
+  def handleEndChargingHelper(currentTick: Int, chargingVehicle: ChargingVehicle): Unit = {
```

Comment:
This needs renamed - I believe it used to be something that was for a more functional style of code where the name made a bit more sense, but now it doesn't. In fact maybe it should just become part of handleEndCharging?

Reply:
Although for readability - maybe renaming is better - processEndChargingEvents?

Reply:
Same questions for the starthelper btw.